src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxTransformNode.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISyntaxTransformNode
+    {
+        ISyntaxTransformBuilder GetBuilder(DriverStateTable previousStateTable);
+    }
+
+    internal interface ISyntaxTransformBuilder
+    {
+        bool Filter(SyntaxNode node);
+
+        void AddFilterFromPreviousTable(SemanticModel model);
+
+        void AddFilterEntries(ImmutableArray<SyntaxNode> nodes, SemanticModel model);
+
+        void SetInputState(DriverStateTable.Builder builder);
+    }
+
+    internal sealed class SyntaxTransformNode<T> : InputNode<T>, ISyntaxTransformNode
+    {
+        private readonly Func<GeneratorSyntaxContext, ImmutableArray<T>> _func;
+
+        private readonly Func<SyntaxNode, bool> _filterFunc;
+
+        private readonly InputNode<SyntaxNode> _filterNode = new InputNode<SyntaxNode>();
+
+        internal SyntaxTransformNode(Func<SyntaxNode, bool> filterFunc, Func<GeneratorSyntaxContext, T> transformFunc)
+        {
+            _func = (t) => ImmutableArray.Create(transformFunc(t));
```

Comment:
>(t) => ImmutableArray.Create(transformFunc(t)) [](http://example.com/codeflow?start=20&length=46)

Could we hold onto `transformFunc` instead and have the one caller create the `ImmutableArray` instead of creating a second delegate (and closure) here.